File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/ActionManagerAssigner.java
package com.spvessel.spacevil;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

import com.spvessel.spacevil.Core.InputEventArgs;
import com.spvessel.spacevil.Flags.InputEventType;

final class ActionManagerAssigner {
    WindowLayout _layout;

    ActionManagerAssigner(WindowLayout layout) {
        _layout = layout;
    }

    void assignActionsForHoveredItem(InputEventType action, InputEventArgs args, Prototype hoveredItem,
            List<Prototype> itemPyramid, boolean isOnlyHovered) {
        if (isOnlyHovered) {
            if (hoveredItem.isDisabled())
                return;

            EventTask task = new EventTask();
            task.item = hoveredItem;
            task.action = action;
            task.args = args;
            _layout.setEventTask(task);
        } else {
            goThroughItemPyramid(itemPyramid, action, args);
        }
        _layout.executePollActions();
    }

    void assignActionsForSender(InputEventType action, InputEventArgs args, Prototype sender,
            List<Prototype> itemPyramid, boolean isPassUnder) {
        if (!sender.isDisabled()) {
            EventTask task = new EventTask();
            task.item = sender;
            task.action = action;
            task.args = args;
            _layout.setEventTask(task);
        }

        if (isPassUnder && sender.isPassEvents(action)) {
            if (itemPyramid != null) {
                goThroughItemPyramid(itemPyramid, action, args);
            }
        }
        _layout.executePollActions();
    }

    void assignActionsForItemPyramid(InputEventType action, InputEventArgs args, Prototype sender,
            List<Prototype> itemPyramid) {
        if (sender.isPassEvents(action)) {
            if (itemPyramid != null) {
                goThroughItemPyramid(itemPyramid, action, args);
            }
        }
        _layout.executePollActions();
    }

    private void goThroughItemPyramid(List<Prototype> itemsList, InputEventType action, InputEventArgs args) {
        Deque<Prototype> tmp = new ArrayDeque<Prototype>(itemsList);
        while (tmp.size() != 0) {

            Prototype item = tmp.pollLast();

            if (item.isDisabled())
                continue;

            EventTask t = new EventTask();
            t.item = item;
            t.action = action;
            t.args = args;
            _layout.setEventTask(t);

            if (!item.isPassEvents(action))
                break;

            // System.out.println(item.getItemName());
        }
    }
}

File: CSharp/SpaceVIL/SpaceVIL/Engine/ActionManagerAssigner.cs
using System;
using System.Collections.Generic;
using SpaceVIL.Core;

namespace SpaceVIL
{
    internal class ActionManagerAssigner
    {
        WindowLayout _layout;

        internal ActionManagerAssigner(WindowLayout layout)
        {
            _layout = layout;
        }

        internal void AssignActionsForHoveredItem(InputEventType action, IInputEventArgs args, Prototype hoveredItem,
                List<Prototype> itemPyramid, bool isOnlyHovered)
        {
            if (isOnlyHovered)
            {
                if (hoveredItem.IsDisabled())
                    return;

                _layout.SetEventTask(new EventTask()
                {
                    Item = hoveredItem,
                    Action = action,
                    Args = args
                });
            }
            else
            {
                GoThroughItemPyramid(itemPyramid, action, args);
            }
            _layout.ExecutePollActions();
        }

        internal void AssignActionsForSender(InputEventType action, IInputEventArgs args, Prototype sender,
                List<Prototype> itemPyramid, bool isPassUnder)
        {
            if (!sender.IsDisabled())
            {
                _layout.SetEventTask(new EventTask()
                {
                    Item = sender,
                    Action = action,
                    Args = args
                });
            }

            if (isPassUnder && sender.IsPassEvents(action))
            {
                if (itemPyramid != null)
                {
                    GoThroughItemPyramid(itemPyramid, action, args);
                }
            }
            _layout.ExecutePollActions();
        }

        internal void AssignActionsForItemPyramid(InputEventType action, IInputEventArgs args, Prototype sender,
                List<Prototype> itemPyramid)
        {
            if (sender.IsPassEvents(action))
            {
                if (itemPyramid != null)
                {
                    GoThroughItemPyramid(itemPyramid, action, args);
                }
            }
            _layout.ExecutePollActions();
        }

        private void GoThroughItemPyramid(List<Prototype> itemsList, InputEventType action, IInputEventArgs args)
        {
            Stack<Prototype> tmp = new Stack<Prototype>(itemsList);
            while (tmp.Count != 0)
            {
                Prototype item = tmp.Pop();

                if (item.IsDisabled())
                    continue;

                _layout.SetEventTask(new EventTask()
                {
                    Item = item,
                    Action = action,
                    Args = args
                });
                
                if (!item.IsPassEvents(action))
                    break;
            }
        }
    }
}